Error when running 'autosubmit create'

In GitLab by @hrosegalb on Feb 22, 2024, 11:07

Hello @dbeltrankyl and @kinow,

### Autosubmit Version
4.0.101

### Summary
I am testing autosubmit locally with a `hello_world` script. When I tried to run `autosubmit create <expid>` to create the experiment, I get the following error:

```
 [CRITICAL] Traceback (most recent call last):
  File "/home/hgalbrai/.pyenv/versions/autosubmit_env/bin/autosubmit", line 37, in main
    Autosubmit.parse_args()
  File "/home/hgalbrai/.pyenv/versions/3.10.12/envs/autosubmit_env/lib/python3.10/site-packages/autosubmit/autosubmit.py", line 659, in parse_args
    Autosubmit._init_logs(args, args.logconsole, args.logfile, expid)
  File "/home/hgalbrai/.pyenv/versions/3.10.12/envs/autosubmit_env/lib/python3.10/site-packages/autosubmit/autosubmit.py", line 801, in _init_logs
    as_conf.reload(force_load=True)
  File "/home/hgalbrai/.pyenv/versions/3.10.12/envs/autosubmit_env/lib/python3.10/site-packages/autosubmitconfigparser/config/configcommon.py", line 1394, in reload
    self.experiment_data = self.normalize_variables(self.experiment_data)
  File "/home/hgalbrai/.pyenv/versions/3.10.12/envs/autosubmit_env/lib/python3.10/site-packages/autosubmitconfigparser/config/configcommon.py", line 508, in normalize_variables
    dependencies = job_data.get("DEPENDENCIES",7000)
AttributeError: 'NoneType' object has no attribute 'get'
'NoneType' object has no attribute 'get'
Unhandled error: If you see this message, please report it in Autosubmit's GitLab project
```


### Steps to reproduce

1. Create a new `expid` using the `Autosubmit-Template` repo. 

```
$ autosubmit expid -H "local" -d "Test in local"
```

2. Update the `expdef_a012.yml` file:

```
DEFAULT:
  # Job experiment ID.
  EXPID: "a012"
  # Default HPC platform name.
  HPCARCH: "local"
EXPERIMENT:
  # List of start dates
  DATELIST: '20000101'
  # List of members.
  MEMBERS: fc0
  # Unit of the chunk size. Can be hour, day, month, or year.
  CHUNKSIZEUNIT: month
  # Size of each chunk.
  CHUNKSIZE: '4'
  # Number of chunks of the experiment.
  NUMCHUNKS: '2'
  CHUNKINI: ''
  # Calendar used for the experiment. Can be standard or noleap.
  CALENDAR: standard
PROJECT:
  # Type of the project.
  PROJECT_TYPE: 'local'
  # Folder to hold the project sources.
  PROJECT_DESTINATION: 'local'
GIT:
  PROJECT_ORIGIN: ''
  PROJECT_BRANCH: ''
  PROJECT_COMMIT: ''
  PROJECT_SUBMODULES: ''
  FETCH_SINGLE_BRANCH: true
SVN:
  PROJECT_URL: ''
  PROJECT_REVISION: ''
LOCAL:
  PROJECT_PATH: '/home/hgalbrai/Projects/autosubmit'
PROJECT_FILES:
  FILE_PROJECT_CONF: ''
  FILE_JOBS_CONF: ''
  JOB_SCRIPTS_TYPE: ''
RERUN:
  RERUN: false
  RERUN_JOBLIST: ''
```

3. Update the `jobs_a012.yml` file:

```
JOBS:
    HELLO_WORLD:
    FILE: "hello_world.sh"
    PLATFORM: local
    RUNNING: once
```

4. Run `autosubmit create a012`

### What is the current bug behavior?

I get the following error message:

```
[ERROR] Trace: 'NoneType' object has no attribute 'get'
 [CRITICAL] Traceback (most recent call last):
  File "/home/hgalbrai/.pyenv/versions/autosubmit_env/bin/autosubmit", line 37, in main
    Autosubmit.parse_args()
  File "/home/hgalbrai/.pyenv/versions/3.10.12/envs/autosubmit_env/lib/python3.10/site-packages/autosubmit/autosubmit.py", line 659, in parse_args
    Autosubmit._init_logs(args, args.logconsole, args.logfile, expid)
  File "/home/hgalbrai/.pyenv/versions/3.10.12/envs/autosubmit_env/lib/python3.10/site-packages/autosubmit/autosubmit.py", line 801, in _init_logs
    as_conf.reload(force_load=True)
  File "/home/hgalbrai/.pyenv/versions/3.10.12/envs/autosubmit_env/lib/python3.10/site-packages/autosubmitconfigparser/config/configcommon.py", line 1394, in reload
    self.experiment_data = self.normalize_variables(self.experiment_data)
  File "/home/hgalbrai/.pyenv/versions/3.10.12/envs/autosubmit_env/lib/python3.10/site-packages/autosubmitconfigparser/config/configcommon.py", line 508, in normalize_variables
    dependencies = job_data.get("DEPENDENCIES",7000)
AttributeError: 'NoneType' object has no attribute 'get'
'NoneType' object has no attribute 'get'
Unhandled error: If you see this message, please report it in Autosubmit's GitLab project
```

### What is the expected correct behavior?

The experiment should be updated with everything needed to run the `hello_world.sh` script and I should not get an error message.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,3 +1,10 @@
+### 4.1.15: Unreleased
+
+**Bug fixes:**
+
+- Corrected the logic for handling `ZombieProcess` errors in `psutil` calls done in `autosubmit stop`, which
+  prevented the command from working correctly if the experiment process appeared after a zombie in the list #2394
+
 ### 4.1.14: Bug fixes, enhancements, and new features
 
 **Bug fixes:**

--- a/autosubmit/helpers/processes.py
+++ b/autosubmit/helpers/processes.py
@@ -146,9 +146,9 @@ def retrieve_expids() -> List[str]:
     expids: List[str] = []
     # NOTE: psutil may raise ``ZombieProcess`` in some cases, even when other exceptions happen in the
     #       case when a process is a zombie (e.g. got a IO error, but the process is zombie)
-    with suppress(ZombieProcess):
-        for process in [p for p in process_iter(['pid', 'cmdline', 'username']) if p.username() == user]:
-            with suppress(ZombieProcess):
+    for process in process_iter(['pid', 'cmdline', 'username']):
+        with suppress(ZombieProcess):
+            if process.username() == user:
                 expid = _match_autosubmit_cmdline(process.cmdline())
                 if expid:
                     expids.append(expid)
@@ -170,16 +170,13 @@ def process_id(expid: str, command="run") -> Optional[int]:
     :rtype: Optional[int]
     """
     user: str = getuser()
-    try:
-        processes = [
-            p for p
-            in process_iter(['pid', 'cmdline', 'username'])
-            if p.username() == user and _match_autosubmit_cmdline(p.cmdline(), command, expid)
-        ]
-    except ZombieProcess:
+    processes = []
+    for process in process_iter(['pid', 'cmdline', 'username']):
         # NOTE: psutil may raise ``ZombieProcess`` in some cases, even when other exceptions happen in the
         #       case when a process is a zombie (e.g. got a IO error, but the process is zombie)
-        return None
+        with suppress(ZombieProcess):
+            if process.username() == user and _match_autosubmit_cmdline(process.cmdline(), command, expid):
+                processes.append(process)
 
     if not processes:
         return None

--- a/test/unit/helpers/test_processes.py
+++ b/test/unit/helpers/test_processes.py
@@ -73,8 +73,11 @@ def test_retrieve_expids(mocker, expids: List[str], expected_retrieved: int, use
     assert len(found) == expected_retrieved
 
 
-def test_retrieve_expids_but_listing_raises_zombies(mocker):
-    mocker.patch('autosubmit.helpers.processes.process_iter', side_effect=ZombieProcess(1984))
+def test_retrieve_expids_but_there_are_zombies_during_listing(mocker):
+    """Test that when we see a ``ZombieProcess`` error during the listing the function returns as expected."""
+    mocked_process = mocker.MagicMock()
+    mocked_process.username.side_effect = ZombieProcess(0)
+    mocker.patch('autosubmit.helpers.processes.process_iter', return_value=[mocked_process])
     expids = retrieve_expids()
     assert type(expids) is list and len(expids) == 0
 
@@ -136,7 +139,9 @@ def test_process_id_multiple_found(mocker):
 
 def test_process_id_but_there_are_zombies(mocker):
     """Test that if the listing of process crashes due to zombies, we get an empty list."""
-    mocker.patch('autosubmit.helpers.processes.process_iter', side_effect=ZombieProcess(1984))
+    mocked_process = mocker.MagicMock()
+    mocked_process.username.side_effect = ZombieProcess(0)
+    mocker.patch('autosubmit.helpers.processes.process_iter', return_value=mocked_process)
 
     pid_found = process_id('a000')
 

